pkg/compiler/codegen.go

```diff
@@ -144,43 +154,90 @@ func (c *codegen) convertBasicType(t types.TypeAndValue, typ *types.Basic) {
 	}
 }
 
-func (c *codegen) emitLoadLocal(name string) {
-	pos := c.scope.loadLocal(name)
-	if pos < 0 {
-		c.prog.Err = fmt.Errorf("cannot load local variable with position: %d", pos)
-		return
-	}
-	c.emitLoadLocalPos(pos)
+func (c *codegen) emitLoadField(i int) {
+	emit.Int(c.prog.BinWriter, int64(i))
+	emit.Opcode(c.prog.BinWriter, opcode.PICKITEM)
 }
 
-func (c *codegen) emitLoadLocalPos(pos int) {
-	emit.Opcode(c.prog.BinWriter, opcode.DUPFROMALTSTACK)
-	emit.Int(c.prog.BinWriter, int64(pos))
-	emit.Opcode(c.prog.BinWriter, opcode.PICKITEM)
+func (c *codegen) emitStoreStructField(i int) {
+	emit.Int(c.prog.BinWriter, int64(i))
+	emit.Opcode(c.prog.BinWriter, opcode.ROT)
+	emit.Opcode(c.prog.BinWriter, opcode.SETITEM)
 }
 
-func (c *codegen) emitStoreLocal(pos int) {
-	emit.Opcode(c.prog.BinWriter, opcode.DUPFROMALTSTACK)
+// getVarIndex returns variable type and position in corresponding slot,
+// according to current scope.
+func (c *codegen) getVarIndex(name string) (varType, int) {
+	if c.scope != nil {
+		if i, ok := c.scope.arguments[name]; ok {
+			return varArgument, i
+		} else if i, ok := c.scope.locals[name]; ok {
+			return varLocal, i
+		}
+	}
+	if i, ok := c.globals[name]; ok {
+		return varGlobal, i
+	}
 
-	if pos < 0 {
-		c.prog.Err = fmt.Errorf("invalid position to store local: %d", pos)
-		return
+	return varLocal, c.scope.newVariable(varLocal, name)
+}
+
+func getBaseOpcode(t varType) (opcode.Opcode, opcode.Opcode) {
+	switch t {
+	case varGlobal:
+		return opcode.LDSFLD0, opcode.STSFLD0
+	case varLocal:
+		return opcode.LDLOC0, opcode.STLOC0
+	case varArgument:
+		return opcode.LDARG0, opcode.STARG0
+	default:
+		panic("invalid type")
 	}
+}
 
-	emit.Int(c.prog.BinWriter, int64(pos))
-	emit.Opcode(c.prog.BinWriter, opcode.ROT)
-	emit.Opcode(c.prog.BinWriter, opcode.SETITEM)
+// emitLoadVar loads specified variable to the evaluation stack.
+func (c *codegen) emitLoadVar(name string) {
+	t, i := c.getVarIndex(name)
+	base, _ := getBaseOpcode(t)
+	if i < 7 {
+		emit.Opcode(c.prog.BinWriter, base+opcode.Opcode(i))
+	} else {
+		emit.Instruction(c.prog.BinWriter, base+7, []byte{byte(i)})
+	}
 }
 
-func (c *codegen) emitLoadField(i int) {
-	emit.Int(c.prog.BinWriter, int64(i))
-	emit.Opcode(c.prog.BinWriter, opcode.PICKITEM)
+// emitStoreVar stores top value from the evaluation stack in the specified variable.
+func (c *codegen) emitStoreVar(name string) {
+	t, i := c.getVarIndex(name)
+	_, base := getBaseOpcode(t)
+	if i < 7 {
+		emit.Opcode(c.prog.BinWriter, base+opcode.Opcode(i))
+	} else {
+		emit.Instruction(c.prog.BinWriter, base+7, []byte{byte(i)})
+	}
 }
 
-func (c *codegen) emitStoreStructField(i int) {
-	emit.Int(c.prog.BinWriter, int64(i))
-	emit.Opcode(c.prog.BinWriter, opcode.ROT)
-	emit.Opcode(c.prog.BinWriter, opcode.SETITEM)
+func (c *codegen) emitDefault(n ast.Expr) {
```

Comment:
Aren't we duplicating some VM logic here? Maybe we could reuse some internal conversions.

Reply:
No, because this `Default` directly relates to `Go` language. VM knows nothing about `Go`.